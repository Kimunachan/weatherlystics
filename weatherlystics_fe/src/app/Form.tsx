"use client";
import { BASE_URL, customStyles } from "@/utils/constants";
import { zodResolver } from "@hookform/resolvers/zod";
import { useMutation, useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useState } from "react";
import { useForm } from "react-hook-form";
import Select from "react-select";
import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import { z } from "zod";
import "../styles/globals.scss";
import styles from "../styles/pages/page.module.scss";

import { WeatherDataType } from "@/utils/types";
import { schema } from "@/utils/schemas";

type FormProps = {
  setWeatherData: (data: WeatherDataType) => void;
};

export default function Form({ setWeatherData }: FormProps) {
  const dateValue = new Date();

  const [showSecondDate, setShowSecondDate] = useState(false);
  const toggleSecondDate = () => setShowSecondDate(!showSecondDate);

  const getWeatherData = useMutation({
    mutationFn: async ({ lat, lon }: { lat: number; lon: number }) => {
      const response = await axios.get(
        `${BASE_URL}/weather?lat=${lat}&lon=${lon}`
      );
      return response.data;
    },
    mutationKey: ["weatherData"],
    onError: (error) => {
      toast.error(`Error fetching weather data: ${error}`);
    },
    onSuccess: (data) => {
      toast.success(`Weather data fetched successfully`);
      setWeatherData(data);
    },
  });

  const [selectedTimezone, setSelectedTimezone] = useState<{
    value: string;
    label: string;
  }>();

  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm<z.infer<typeof schema>>({
    resolver: zodResolver(schema),
    mode: "onBlur",
    reValidateMode: "onBlur",
    defaultValues: {
      lat: 0,
      long: 0,
    },
  });

  const onSubmit = handleSubmit((data) => {
    getWeatherData.mutate({
      lat: data.lat,
      lon: data.long,
    });
  });

  const useMyLocation = async () => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition((position) => {
        const { latitude, longitude } = position.coords;
        setValue("lat", latitude);
        setValue("long", longitude);
      });
    } else {
      toast.error("Geolocation is not supported by this browser.");
    }
  };

  const {
    data: timezones,
    isLoading: isLoadingTimezones,
    isError,
  } = useQuery({
    queryKey: ["timezones"],
    queryFn: async () => {
      const response = await axios.get<string[]>(
        "http://worldtimeapi.org/api/timezone"
      );
      return response.data.map((timezone) => ({
        value: timezone,
        label: timezone,
      }));
    },
  });

  const handleTimezoneChange = (
    selectedOption: { value: string; label: string } | null
  ) => {
    if (selectedOption) {
      setSelectedTimezone(selectedOption);
      setValue("timezone", selectedOption.value);
    }
  };
  const handleLatChange = (event: any) => {
    setValue("lat", event.target.value);
  };

  const handleLongChange = (event: any) => {
    setValue("long", event.target.value);
  };

  if (isLoadingTimezones) return <div>Loading...</div>;
  if (isError) toast.error("Error fetching timezones");

  return (
    <>
      <form className="form" onSubmit={onSubmit}>
        <div className="form-row">
          <label>
            Latitude:
            <input
              type="number"
              step={0.0000001}
              {...register("lat", { required: true, valueAsNumber: true })}
              onChange={handleLatChange}
            />
            {errors.lat && <p>{errors.lat.message}</p>}
          </label>
          <label>
            Longitude:
            <input
              type="number"
              step={0.0000001}
              {...register("long", { required: true, valueAsNumber: true })}
              onChange={handleLongChange}
            />
            {errors.long && <p>{errors.long.message}</p>}
          </label>
        </div>
        <div className="form-row">
          <label>
            Timezone:
            <Select
              isLoading={isLoadingTimezones}
              options={timezones}
              styles={customStyles}
              onChange={handleTimezoneChange}
              value={selectedTimezone}
            />
            {errors.timezone && <p>{errors.timezone.message}</p>}
          </label>
          <button
            type="button"
            className={styles.normalButton}
            onClick={useMyLocation}
          >
            Use my location
          </button>
        </div>
        <div className="form-row">
          <label>
            Date:
            <input
              type="date"
              defaultValue={dateValue.toISOString().split("T")[0]}
              {...register("date", {
                required: true,
                valueAsDate: true,
              })}
            />
            {errors.date && <p>{errors.date.message}</p>}
          </label>
          {showSecondDate && (
            <label>
              Second Date
              <input
                type="date"
                defaultValue={dateValue.toISOString().split("T")[0]}
                {...register("secondDate", { valueAsDate: true })}
              />
              {errors.secondDate && <p>{errors.secondDate.message}</p>}
            </label>
          )}
          <button className={styles.circleButton} onClick={toggleSecondDate}>
            {showSecondDate ? "-" : "add"}
          </button>
        </div>
        <button className={styles.longButton} type="submit">
          Submit
        </button>
      </form>
    </>
  );
}
