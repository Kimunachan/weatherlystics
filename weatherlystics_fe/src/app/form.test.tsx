import {
    fireEvent,
    render,
    screen,
    waitFor,
    waitForElementToBeRemoved,
} from "@/utils/customTestUtils";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { toast } from "react-toastify";
import Form from "./Form";
  
  jest.mock("react-toastify", () => ({
    toast: {
      error: jest.fn(),
      success: jest.fn(),
    },
  }));
  
  const axiosMock = new MockAdapter(axios);
  
  describe("Form Component", () => {
    const mockSetWeatherData = jest.fn();
  
    beforeEach(() => {
      axiosMock.reset();
    });
  
    it("should display error messages for all invalid fields", async () => {
        axiosMock
            .onGet("http://worldtimeapi.org/api/timezone")
            .replyOnce(200, ["Europe/Berlin", "America/New_York"]);
    
        render(<Form setWeatherData={mockSetWeatherData} />);
    
        await waitForElementToBeRemoved(() => screen.getByText("Loading..."));
    
        fireEvent.click(screen.getByText('Submit'));
    
        await waitFor(() => {
            const alerts = screen.getAllByRole('alert');
            expect(alerts).toHaveLength(1); // lat, long, and date
            alerts.forEach(alert => {
            expect(alert).toHaveTextContent('Required');
            });
        });
    });

    it("should call toast.error if there is an error fetching weather data", async () => {
        const mockSetWeatherData = jest.fn();
    
        axiosMock
            .onGet("http://worldtimeapi.org/api/timezone")
            .replyOnce(200, ["Europe/Berlin", "America/New_York"]);
    
        axiosMock.onGet(/weather/).replyOnce(500, { message: "Error" });
    
        render(<Form setWeatherData={mockSetWeatherData} />);
    
        await waitForElementToBeRemoved(() => screen.getByText("Loading..."));
    
        fireEvent.change(screen.getByLabelText("Latitude:"), {
            target: { value: 52.52 },
        });
        fireEvent.change(screen.getByLabelText("Longitude:"), {
            target: { value: 13.405 },
        });
    
        const selectInput = screen.getByRole("combobox");
        fireEvent.keyDown(selectInput, { key: "ArrowDown", code: "ArrowDown" });
    
        await waitFor(() => screen.getByText("Europe/Berlin"));
    
        fireEvent.click(screen.getByText("Europe/Berlin"));
    
        fireEvent.submit(screen.getByTestId("form"));
    
        await waitFor(() => {
            expect(toast.error).toHaveBeenCalledWith("Error fetching weather data: Error: Request failed with status code 500");
        });
      });
  
      it("should call toast.error if geolocation is not supported",async () => {
        const mockSetWeatherData = jest.fn();
    
        axiosMock
          .onGet("http://worldtimeapi.org/api/timezone")
          .replyOnce(200, ["Europe/Berlin", "America/New_York"]);
    
        render(<Form setWeatherData={mockSetWeatherData} />);
    
        await waitForElementToBeRemoved(() => screen.getByText("Loading..."));
    
        Object.defineProperty(global.navigator, "geolocation", {
          value: undefined,
          configurable: true,
        });
    
        fireEvent.click(screen.getByText("Use my location"));
    
        expect(toast.error).toHaveBeenCalledWith("Geolocation is not supported by this browser.");
      });
  
    it("should toggle second date input", async () => {
        axiosMock
            .onGet("http://worldtimeapi.org/api/timezone")
            .replyOnce(200, ["Europe/Berlin", "America/New_York"]);
    
        render(<Form setWeatherData={mockSetWeatherData} />);
    
        await waitForElementToBeRemoved(() => screen.getByText("Loading..."));
    
        fireEvent.click(screen.getByText("add"));
    
        expect(screen.getByLabelText(/Second Date/i)).toBeInTheDocument();
    
        fireEvent.click(screen.getByText("add"));
    
        expect(screen.queryByLabelText(/Second Date/i)).not.toBeInTheDocument();
    });
  
    it("should handle timezone selection", async () => {
        const mockSetWeatherData = jest.fn();

        axiosMock
        .onGet("http://worldtimeapi.org/api/timezone")
        .replyOnce(200, ["Europe/Berlin", "America/New_York"]);

        axiosMock.onGet(/weather/).replyOnce(500, { message: "Error" });

        render(<Form setWeatherData={mockSetWeatherData} />);

        await waitForElementToBeRemoved(() => screen.getByText("Loading..."));

        const selectInput = screen.getByRole("combobox");
            fireEvent.keyDown(selectInput, { key: "ArrowDown", code: "ArrowDown" });

            await waitFor(() => screen.getByText("Europe/Berlin"));

            fireEvent.click(screen.getByText("Europe/Berlin"));

        await waitFor(() => {
            expect(screen.getByText("Europe/Berlin")).toBeInTheDocument();
        });
    });
  
    it("should show loading state for timezones", async () => {
        axiosMock.onGet("http://worldtimeapi.org/api/timezone").reply(200, []);
    
        render(<Form setWeatherData={mockSetWeatherData} />);
    
        expect(screen.getByText("Loading...")).toBeInTheDocument();
    });
    
    it("should render the form and interact correctly", async () => {
        const mockSetWeatherData = jest.fn();
    
        axiosMock
        .onGet("http://worldtimeapi.org/api/timezone")
        .replyOnce(200, ["Europe/Berlin", "America/New_York"]);
    
        axiosMock
        .onGet(/weather/)
        .replyOnce(200, { data: "mockWeatherData" });
    
        render(<Form setWeatherData={mockSetWeatherData} />);
    
        await waitForElementToBeRemoved(() => screen.getByText("Loading..."));
    
        fireEvent.change(screen.getByLabelText("Latitude:"), {
            target: { value: 52.52 },
        });
        fireEvent.change(screen.getByLabelText("Longitude:"), {
            target: { value: 13.405 },
        });
    
        const selectInput = screen.getByRole("combobox");
        fireEvent.keyDown(selectInput, { key: "ArrowDown", code: "ArrowDown" });
    
        await waitFor(() => screen.getByText("Europe/Berlin"));
    
        fireEvent.click(screen.getByText("Europe/Berlin"));
    
        fireEvent.submit(screen.getByTestId("form"));
    
        await waitFor(() => {
            expect(screen.getByTestId("form")).toBeInTheDocument();
            expect(screen.getByLabelText("Latitude:")).toHaveValue(52.52);
            expect(screen.getByLabelText("Longitude:")).toHaveValue(13.405);
            expect(screen.getByLabelText("Date:")).toHaveValue(
            new Date('2024-05-22T00:00:00Z').toISOString().slice(0, 10)
            );
        });
    
        expect(mockSetWeatherData).toHaveBeenCalled();
        const form = screen.getByTestId("form");
        expect(form).toMatchSnapshot();
    });

    it("should handle 'Use my location' button click", async () => {
        const mockSetWeatherData = jest.fn();
    
        axiosMock
        .onGet("http://worldtimeapi.org/api/timezone")
        .replyOnce(200, ["Europe/Berlin", "America/New_York"]);
    
        render(<Form setWeatherData={mockSetWeatherData} />);
    
        await waitForElementToBeRemoved(() => screen.getByText("Loading..."));
    
        const mockGeolocation = {
            getCurrentPosition: jest.fn().mockImplementation((success) =>
                success({
                    coords: {
                        latitude: 52.52,
                        longitude: 13.405,
                    },
                })
            ),
        };
    
        // Mock navigator.geolocation with jest-location-mock
        Object.defineProperty(global.navigator, 'geolocation', {
            value: mockGeolocation,
            configurable: true,
        });
    
        fireEvent.click(screen.getByText("Use my location"));
    
        await waitFor(() => {
            expect(screen.getByLabelText("Latitude:")).toHaveValue(52.52);
            expect(screen.getByLabelText("Longitude:")).toHaveValue(13.405);
        });
    });
});
