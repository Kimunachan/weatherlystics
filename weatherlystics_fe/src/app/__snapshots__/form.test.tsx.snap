// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component should render the form and interact correctly 1`] = `
<form
  class="form"
  data-testid="form"
>
  <div
    class="form-row"
    data-testid="row"
  >
    <label
      data-testid="Latitude"
    >
      Latitude:
      <input
        name="rows.0.lat"
        step="1e-7"
        type="number"
      />
    </label>
    <label
      data-testid="Longitude"
    >
      Longitude:
      <input
        name="rows.0.long"
        step="1e-7"
        type="number"
      />
    </label>
    <label
      data-testid="Timezone"
    >
      Timezone:
      <div
        class=" css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-6-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class=" css-er85fr-control"
        >
          <div
            class=" css-1fdsijx-ValueContainer"
          >
            <div
              class=" css-yn97rf-singleValue"
            >
              Europe/Berlin
            </div>
            <div
              class=" css-1ji2ryk-Input"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-6-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 0; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-yl5mvd-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="rows.0.timezone"
          type="hidden"
          value="Europe/Berlin"
        />
      </div>
    </label>
    <label>
      Date:
      <input
        max="2024-06-09"
        min="1940-01-01"
        name="rows.0.date"
        type="date"
        value="2024-06-01"
      />
    </label>
    <button
      class="normalButton"
      type="button"
    >
      Use my location
    </button>
    <button
      class="circleButton"
      data-testid="add-0"
      type="button"
    >
      +
    </button>
  </div>
  <button
    class="longButton"
    type="submit"
  >
    Submit
  </button>
</form>
`;
